Add WatchYourLAN integration

drwahl opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
Adding support for WatchYourLAN (https://github.com/aceberg/WatchYourLAN) integration. This will allow HA to fetch information from the WatchYourLAN API and process the results at entities.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]
https://github.com/home-assistant/home-assistant.io/pull/34584

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

drwahl committed:
feat: Add WatchYourLAN integration

drwahl committed:
Update update interval to 5 seconds

gravytrain commented:
Approved +2

joostlek commented on homeassistant/components/watchyourlan/__init__.py:
```diff
@@ -0,0 +1,38 @@
+    update_interval = entry.options.get("update_interval", UPDATE_FREQUENCY)
+
+    # Store the config entry data in hass.data for use by platforms
+    hass.data[DOMAIN][entry.entry_id] = {
+        "host": entry.data[CONF_HOST],
+        "port": entry.data[CONF_PORT],
+        "ssl": entry.data[CONF_SSL],
+        "url": entry.data["url"],
+        "update_interval": update_interval,  # Use the user-configured interval or default
+    }
```
All this data is already available in the entry and thus is not relevant to store in hass.data

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,71 @@
+        vol.Required(CONF_HOST): str,
+        vol.Optional(CONF_PORT, default=8840): int,
+        vol.Optional(CONF_SSL, default=False): bool,
+        vol.Optional("update_interval", default=5): vol.All(
+            vol.Coerce(int), vol.Range(min=1, max=60)
+        ),
```
update interval is not allowed to be user configurable. Users can use the `homassistant.update_entity` service to do that

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,71 @@
+)
+
+
+async def validate_input(hass, data):
+    """Validate the user input allows us to connect to the API."""
+    proto = "https" if data[CONF_SSL] else "http"
+    target = f"{proto}://{data[CONF_HOST]}:{data[CONF_PORT]}"
+
+    async with get_async_client(hass) as client:
+        response = await client.get(f"{target}/api/all")
+        if response.status_code != 200:
+            raise CannotConnect
+    return {"title": "WatchYourLAN", "url": target}
```
All logic that is required to connect to a device or service should be put in a library hosted on pypi

drwahl committed:
feat: Update WatchYourLAN integration to remove unused code and improve performance

drwahl commented:
@joostlek I appreciate the thorough review and think I've been able to address all the comments. Looking forward to another round of review, when you have time!

drwahl committed:
Update WatchYourLAN integration to use the latest version of the watchyourlanclient library

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,60 @@
+)
+
+
+async def validate_input(hass, data):
```
This is tiny, can be inlined

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,60 @@
+    return {"title": "WatchYourLAN", "url": target}
+
+
+class WatchYourLANConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```
```suggestion
class WatchYourLANConfigFlow(ConfigFlow, domain=DOMAIN):
```

joostlek commented on homeassistant/components/watchyourlan/coordinator.py:
```diff
@@ -0,0 +1,37 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class WatchYourLANUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching data from the WatchYourLAN API."""
+
+    def __init__(self, hass: HomeAssistant, config: MappingProxyType) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="WatchYourLAN",
+        )
+        self.api_url = config.get("url")
+        self.api_client = WatchYourLANClient(
+            base_url=config.get("url"), async_mode=True
+        )
+
+    async def _async_update_data(self):
+        """Fetch data from the WatchYourLAN API with retries."""
+        try:
+            return await self.api_client.get_all_hosts()
+        except Exception as e:
```
This coordinator does not poll regularly, why?

joostlek commented on homeassistant/components/watchyourlan/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "name": "WatchYourLAN",
+  "codeowners": ["@drwahl"],
+  "config_flow": true,
+  "dependencies": [],
+  "documentation": "https://www.home-assistant.io/integrations/watchyourlan",
+  "homekit": {},
+  "iot_class": "local_polling",
+  "requirements": ["watchyourlanclient==2024.1009a0"],
+  "ssdp": [],
+  "zeroconf": []
```
Remove empty fields

joostlek commented on homeassistant/components/watchyourlan/__init__.py:
```diff
@@ -0,0 +1,21 @@
+from .const import DOMAIN
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up WatchYourLAN from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+
+    await hass.config_entries.async_forward_entry_setup(entry, Platform.SENSOR)
+
+    return True
+
```
Can we use the init to check if the integration is able to setup?

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,134 @@
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the WatchYourLAN sensors."""
+    coordinator = WatchYourLANUpdateCoordinator(hass, entry.data)
+    await coordinator.async_config_entry_first_refresh()
```
this shound be moved to the init. The coordinator can be stored in entry.runtime_data

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,134 @@
+
+    entities = [WatchYourLANDeviceCountSensor(coordinator)]
+
+    if isinstance(coordinator.data, list):
```
This is always the case

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,134 @@
+        self.device = device
+        self.base_url = config["url"]
+
+    @property
+    def name(self):
+        """Return the name of the sensor, using the format 'WatchYourLAN <ID> <Optionally include name>'."""
+        device_id = self.device.get("ID")
+        device_name = self.device.get("Name", "").strip()
+
+        if not device_name:
+            return f"WatchYourLAN {device_id}"
+        return f"WatchYourLAN {device_id} {device_name}"
```
Please use `_attr_has_entity_name = True` and add an `_attr_device_info`

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,134 @@
+        """Return the state of the sensor."""
+        return "Online" if self.device.get("Now") == 1 else "Offline"
+
+    @property
+    def extra_state_attributes(self):
+        """Return other attributes related to the device, including the link to the device's page."""
+        host_id = self.device.get("ID")
+        host_page_url = f"{self.base_url}/host/{host_id}"
+
+        return {
+            "ID": self.device.get("ID"),
+            "Name": self.device.get("Name"),
+            "IP": self.device.get("IP"),
+            "Mac": self.device.get("Mac"),
+            "Hardware": self.device.get("Hw"),
+            "Iface": self.device.get("Iface"),
+            "DNS": self.device.get("DNS"),
+            "Last_Seen": self.device.get("Date"),
+            "Known": self.device.get("Known"),
+            "Now": self.device.get("Now"),
+            "Host Page": host_page_url,
+        }
```
A lot of these datapoints can be their own sensor

drwahl replied:
Is it preferred to have them be their own sensors, or attributes of a single sensor? I can certainly make them individual sensors if that is the preferred approach, but that seemed messy because it created so many sensors.

joostlek replied:
We want entities rather than attributes. They're more efficient for the database and more customizable for the user. Also, some of these fields should be stored via the device info

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,134 @@
+            "Host Page": host_page_url,
+        }
+
+    @property
+    def available(self):
+        """Return if the sensor is available."""
+        return self.coordinator.last_update_success
```
Pleae implement a CoordinatorEntity as well. This is the default implementation

joostlek reviewed (changes requested):
The integration needs more work to be clean and neat. Checkout an integation like `airgradient` for a good example to look to

drwahl committed:
feat: Update WatchYourLAN integration to use latest watchyourlanclient library version

drwahl committed:
feat: Update WatchYourLAN integration to use latest watchyourlanclient library version

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,247 @@
+    async_add_entities(entities)
+
+
+class WatchYourLANSensor(CoordinatorEntity, RestoreSensor, SensorEntity):
+    """Representation of a WatchYourLAN online/offline sensor."""
+
+    def __init__(
+        self, coordinator: WatchYourLANUpdateCoordinator, device: dict
+    ) -> None:
+        """Initialize the sensor for online/offline state."""
+        super().__init__(coordinator)
+        self.device = device
+        self.entity_description = ONLINE_STATUS_DESCRIPTION
+
+        # Set the unique ID and device info
+        self._attr_unique_id = f"{self.device.get('ID')}_online_status"
+        self._attr_device_info = build_device_info(self.device, DOMAIN)
+
+        # Set an appropriate icon for the sensor
+        self._attr_icon = (
+            "mdi:lan-connect"
+            if self.device.get("Now", 0) == 1
+            else "mdi:lan-disconnect"
+        )
+
+    @property
+    def native_value(self) -> str:
+        """Return the native value of the sensor."""
+        return "Online" if self.device.get("Now", 0) == 1 else "Offline"
+
+    async def async_update(self):
+        """Update the sensor's state from the coordinator."""
+        await (
+            self.coordinator.async_request_refresh()
+        )  # Request data update from the coordinator
+        self._attr_native_value = (
+            "Online" if self.device.get("Now", 0) == 1 else "Offline"
+        )
```
The idea of entity_descriptions is that you create 1 generic base entity, and then use entity_descriptions to change how they look and what data to expose

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,247 @@
+    async_add_entities(entities)
+
+
+class WatchYourLANSensor(CoordinatorEntity, RestoreSensor, SensorEntity):
```
```suggestion
class WatchYourLANSensor(CoordinatorEntity[..], RestoreSensor, SensorEntity):
```
Please add the coordinator type in the CoordinatorEntity

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,247 @@
+    async_add_entities(entities)
+
+
+class WatchYourLANSensor(CoordinatorEntity, RestoreSensor, SensorEntity):
```
Why the RestoreSensor?

drwahl replied:
It probably isn't needed. I was tinkering with optimizing things on reload/startup, but I think it's going to reliably hit the API anyway, so the RestoreSensor probably isn't really needed.

drwahl commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,63 @@
+                api_client = WatchYourLANClient(base_url=target, async_mode=True)
+                hosts = await api_client.get_all_hosts()
+                if not hosts:
+                    raise CannotConnect  # noqa: TRY301
```
It's complaining about raising an error inside a `try`. However, this seems like a time that really does make it more user-friendly to raise the "CannotConnect" instead of whatever the api_client might throw.

joostlek replied:
You can set errors, and then change the `else` of the `try` block to `if not errors`

drwahl commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,63 @@
+
+                # Return a config entry on successful connection
+                return self.async_create_entry(
+                    title="WatchYourLAN", data={"url": target}
```
Agreed. That will make this a bit simpler, too.

drwahl commented on homeassistant/components/watchyourlan/coordinator.py:
```diff
@@ -0,0 +1,40 @@
+            base_url=config.get("url"), async_mode=True
+        )
+
+    async def _async_update_data(self):
+        """Fetch data from the WatchYourLAN API with retries."""
+        try:
+            return await self.api_client.get_all_hosts()
```
I can add some basic typing (list of dicts), but that's about the best I can do at the moment.

drwahl committed:
Update WatchYourLAN integration to use latest watchyourlanclient library version

drwahl committed:
Update WatchYourLAN integration to use latest watchyourlanclient library version

joostlek commented on homeassistant/components/watchyourlan/__init__.py:
```diff
@@ -0,0 +1,34 @@
+
+from .coordinator import WatchYourLANUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
```
Unused

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,63 @@
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_URL): str,
+        vol.Optional(CONF_SSL, default=False): bool,
```
Why do we have this, didn't seem like we use it

drwahl replied:
Good catch. I forgot to pass this through to the WatchYourLAN client.

joostlek commented on homeassistant/components/watchyourlan/coordinator.py:
```diff
@@ -0,0 +1,41 @@
+class WatchYourLANUpdateCoordinator(DataUpdateCoordinator[list[dict[str, Any]]]):
+    """Class to manage fetching data from the WatchYourLAN API."""
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
```
```suggestion
    config_entry: ConfigEntry


    def __init__(self, hass: HomeAssistant) -> None:
```

joostlek commented on homeassistant/components/watchyourlan/coordinator.py:
```diff
@@ -0,0 +1,41 @@
+            name="WatchYourLAN",
+            update_interval=update_interval,  # Add polling interval here
+        )
+        self.api_url = config_entry.data[CONF_URL]
```
```suggestion
        self.api_url = self.config_entry.data[CONF_URL]
```
Where do we use this variable more?

drwahl replied:
It's only used in the api_client setup. No need for it to be here, really. I'll remove it.

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,155 @@
+        return field_mapping.get(self.entity_description.key, "")
+
+
+class WatchYourLANDeviceCountSensor(
```
Let's split this off to a second PR, this way we can focus on the first part and add the rest later

joostlek replied:
Still relevant

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,155 @@
+
+
+# Utility function for setting up device info consistently
+def build_device_info(device: dict[str, Any], domain: str) -> DeviceInfo:
```
Move this to the entity itself

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,155 @@
+        name=device.get("Name") or f"WatchYourLAN {device.get('ID', 'Unknown')}",
+        manufacturer=device.get("Hw", "Unknown Manufacturer"),
+        model="WatchYourLAN Device",
+        sw_version=device.get("Last_Seen", "Unknown"),
```
Is last seen the version?

drwahl replied:
Not really, but I wasn't sure what else to use for this. There really is no "version". Would it be best to just leave sw_version unset?

joostlek replied:
It's not required, so leave it unset if you don't have it

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,155 @@
+# Utility function for setting up device info consistently
+def build_device_info(device: dict[str, Any], domain: str) -> DeviceInfo:
+    """Build device info helper function."""
+    return DeviceInfo(
```
Do we have the Mac?

drwahl replied:
As far as I know, we should always have the MAC. Otherwise, I'm not sure how WatchYourLAN is able to ping the IP.

joostlek replied:
Let's use `connections` instead of identifiers, and have the Mac address registered

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,155 @@
+
+    # Loop over each device and create sensors for each entity description
+    for device in coordinator.data:
+        if isinstance(device, dict):
```
Why would it not be a dict?

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,155 @@
+                for description in ENTITY_DESCRIPTIONS
+            ]
+
+    async_add_entities(entities)
```
Create a single list comprehension from the for loops

joostlek commented on homeassistant/components/watchyourlan/strings.json:
```diff
@@ -0,0 +1,21 @@
+{
```
Strings are stale

drwahl committed:
Update WatchYourLAN integration error handling

drwahl committed:
feat: Update WatchYourLAN integration to use latest watchyourlanclient library version

drwahl committed:
Refactor WatchYourLAN sensor setup to use list comprehension

drwahl committed:
chore: Remove unused code in WatchYourLANGenericSensor class

drwahl committed:
chore: Handle connection errors and exceptions in WatchYourLAN setup

drwahl committed:
chore: Update WatchYourLAN sensor device info to use MAC address as connection

autinerd commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_URL): str,
+        vol.Optional(CONF_SSL, default=False): bool,
```
```suggestion
        vol.Optional(CONF_VERIFY_SSL, default=False): bool,
```
As the parameter of the client is called `verify_ssl`, the desired config seems to be `CONF_VERIFY_SSL`

autinerd commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+from watchyourlanclient import WatchYourLANClient
+
+from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
+from homeassistant.const import CONF_SSL, CONF_URL
```
```suggestion
from homeassistant.const import CONF_URL, CONF_VERIFY_SSL
```

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+                    verify_ssl=user_input[CONF_SSL],
+                )
+                hosts = await api_client.get_all_hosts()
+                if not hosts:
+                    errors["base"] = "cannot_connect"
+            except (ConnectError, HTTPStatusError) as exc:
+                _LOGGER.error("Connection error during setup: %s", exc)
+                errors["base"] = "cannot_connect"
+            except Exception:  # noqa: BLE001
+                errors["base"] = "unknown"
+            else:
+                # Return a config entry on successful connection
+                if not errors:
+                    return self.async_create_entry(
+                        title="WatchYourLAN", data={"url": user_input[CONF_URL]}
+                    )
```
```suggestion
            except (ConnectError, HTTPStatusError) as exc:
                _LOGGER.error("Connection error during setup: %s", exc)
                errors["base"] = "cannot_connect"
            except Exception:  # noqa: BLE001
                errors["base"] = "unknown"
            if not hosts:
                errors["base"] = "cannot_connect"
            if not errors:
                return self.async_create_entry(
                    title="WatchYourLAN", data={"url": user_input[CONF_URL]}
                )
```

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+                # Return a config entry on successful connection
+                if not errors:
+                    return self.async_create_entry(
+                        title="WatchYourLAN", data={"url": user_input[CONF_URL]}
```
Why don't we store the SSL setting?

joostlek commented on homeassistant/components/watchyourlan/coordinator.py:
```diff
@@ -0,0 +1,40 @@
+
+
+class WatchYourLANUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching data from the WatchYourLAN API."""
```
Please just do this change 😂

joostlek commented on homeassistant/components/watchyourlan/coordinator.py:
```diff
@@ -0,0 +1,47 @@
+            update_interval=update_interval,
+        )
+
+        # Ensure self.config_entry is not None before accessing its data
+        if self.config_entry is None:
+            raise ValueError("config_entry is None, cannot access configuration data")
```
Because this is not needed

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,149 @@
+from .coordinator import WatchYourLANUpdateCoordinator
+
+# Define entity descriptions for each sensor type
+ENTITY_DESCRIPTIONS = [
+    SensorEntityDescription(
+        key="online_status",
+        name="Online Status",
+        icon="mdi:lan-connect",
+    ),
+    SensorEntityDescription(
+        key="ip_address",
+        name="IP Address",
+        icon="mdi:ip-network",
+    ),
+    SensorEntityDescription(
+        key="mac_address",
+        name="MAC Address",
+        icon="mdi:lan",
+    ),
+    SensorEntityDescription(
+        key="iface",
+        name="Network Interface",
+        icon="mdi:ethernet",
+    ),
+]
```
Please use translations and icon translations

drwahl replied:
I got the icon tanslation working, but not the name translation. Do you have any pointers for where I'm going wrong here?

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,149 @@
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
```
Use the extended ConfigEntry type

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,149 @@
+        super().__init__(coordinator)
+        self.device = device
+        self.entity_description = description
+        self._attr_unique_id = f"{self.device.get('ID')}_{description.key}"
```
Why don't we use mac address?

drwahl replied:
The schema in WatchYourLAN uses "ID" as the primary key and "mac" doesn't *have* to be unique. Also, if the WatchYourLAN DB gets wiped out, it's possible a previous ID now maps to a different entity within WatchYourLAN or may not exist at all anymore. It seems more inline with how WatchYourLAN works to base the unique_id on the "ID" field instead of the mac addr.

joostlek replied:
> it's possible a previous ID now maps to a different entity within WatchYourLAN or may not exist at all anymore

That kinda sounds like a reason to use the mac address. Yes, maybe the same mac address could appear in the data twice, but they are still used for the same device

joostlek replied:
Or am I missing something here?

joostlek replied:
I still have this question by the way

drwahl replied:
Oh, weird, I thought I committed my change to switch this to the Mac Addr. I think you make a good point here regarding the Mac Addr being the definitive/unique thing, and it makes sense to use it for the device unique ID.

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,149 @@
+    @property
+    def native_value(self) -> str:
+        """Return the native value of the sensor based on its description."""
+        if self.entity_description.key == "online_status":
+            return "Online" if self.device.get("Now", 0) == 1 else "Offline"
```
That's a binary sensor, not a normal sensor

joostlek replied:
Still relevant

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,149 @@
+        }
+        return field_mapping.get(self.entity_description.key, "")
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return device info for the sensor."""
+        return DeviceInfo(
+            connections={
```
move to the constructor with `self._attr_device_info = ...`

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,149 @@
+        """Return device info for the sensor."""
+        return DeviceInfo(
+            connections={
+                (CONNECTION_NETWORK_MAC, self.device.get("Mac", "00:00:00:00:00:00"))
```
Didn't you say you always had a mac?

drwahl replied:
We do/should, but I don't always trust APIs :) I figured it was just safer this way, instead of throwing some IndexError or something.

joostlek replied:
If the API docs say that you have an mac address, just go for it. Once people report it doesn't work like expected, we can always implement a fix. I'd rather do this than have code that gives the impression it will be optional

drwahl committed:
chore: Update WatchYourLAN sensor device info to use MAC address as connection

drwahl committed:
Update homeassistant/components/watchyourlan/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

drwahl committed:
chore: Update WatchYourLAN sensor device info to use MAC address as connection

drwahl committed:
Update homeassistant/components/watchyourlan/coordinator.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

drwahl committed:
chore: Update WatchYourLAN sensor device info to use MAC address as connection

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,140 @@
+        super().__init__(coordinator)
+        self.device = device
+        self._attr_unique_id = f"{self.device.get('Mac')}_online_status"
+        mac_address = self.device.get("Mac", "Unknown")
```
I commented on the other MAC topic

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,140 @@
+        key="ip_address",
+        translation_key="ip_address",
+    ),
+    SensorEntityDescription(
+        key="mac_address",
+        translation_key="mac_address",
+    ),
```
Since the device_info already has the mac, we don't put it as a sensor here

joostlek reviewed (changes requested):
I'll do a full review pass tomorrow morning

drwahl committed:
chore: Update WatchYourLAN sensor device info to use MAC address as connection

drwahl committed:
chore: Update WatchYourLAN sensor device info to use MAC address as connection

drwahl committed:
chore: Remove mac_address from WatchYourLAN sensor entity descriptions

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+        errors = {}
+        if user_input is not None:
+            try:
+                # Use the WatchYourLANClient to validate the connection
+                api_client = WatchYourLANClient(
+                    base_url=user_input[CONF_URL],
+                    async_mode=True,
+                    verify_ssl=user_input[CONF_VERIFY_SSL],
+                )
```
Does creating the object raise?

drwahl replied:
No, creating the client does raise. The client only raises when the call(s) to the API actually fail, which requires calling the client's methods.

joostlek replied:
Okay so let's move it out of the try block

joostlek commented on homeassistant/components/watchyourlan/icons.json:
```diff
@@ -0,0 +1,18 @@
+      "ip_address": {
+        "default": "mdi:ip-network"
+      },
+      "mac_address": {
+        "default": "mdi:lan"
+      },
```
unused

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,136 @@
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
```
Use the extended ConfigEntry type

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,136 @@
+        self._attr_unique_id = f"{self.device.get('ID')}_{description.key}"
+        self._attr_device_info = DeviceInfo(
+            connections={(CONNECTION_NETWORK_MAC, self.device["Mac"])},
+            name=self.device.get("Name")
+            or f"WatchYourLAN {self.device.get('ID', 'Unknown')}",
```
Please wrap this in parenthesis

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,136 @@
+        self._attr_unique_id = f"{self.device.get('ID')}_{description.key}"
+        self._attr_device_info = DeviceInfo(
+            connections={(CONNECTION_NETWORK_MAC, self.device["Mac"])},
+            name=self.device.get("Name")
+            or f"WatchYourLAN {self.device.get('ID', 'Unknown')}",
+            manufacturer=self.device.get("Hw", "Unknown Manufacturer"),
```
```suggestion
            manufacturer=self.device.get("Hw"),
```
Let's not add one if its unknown

joostlek replied:
```suggestion
            default_name=self.device.get("Name")
            or f"WatchYourLAN {self.device.get('ID', 'Unknown')}",
            default_manufacturer=self.device.get("Hw", "Unknown Manufacturer"),
```
This way if one of the devices is a device that is integrated via HA, that name takes priority over this one

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,136 @@
+            name=self.device.get("Name")
+            or f"WatchYourLAN {self.device.get('ID', 'Unknown')}",
+            manufacturer=self.device.get("Hw", "Unknown Manufacturer"),
+            model="WatchYourLAN Device",
```
We can remove this one since it's not a normal model

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,136 @@
+        """Return the native value of the sensor based on its description."""
+        if self.entity_description.key == "online_status":
+            return "Online" if self.device.get("Now", 0) == 1 else "Offline"
+        return self.device.get(self._get_device_field_for_key(), "Unknown")
```
To make an entity return `Unknown`, return `None` here

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,136 @@
+            return "Online" if self.device.get("Now", 0) == 1 else "Offline"
+        return self.device.get(self._get_device_field_for_key(), "Unknown")
+
+    def _get_device_field_for_key(self) -> str:
+        """Map description key to the appropriate device field."""
+        field_mapping = {
+            "online_status": "Now",
+            "ip_address": "IP",
+            "mac_address": "Mac",
+            "iface": "Iface",
+        }
+        return field_mapping.get(self.entity_description.key, "")
```
Rather extend the EntityDescription to add a second key to use to get the data

drwahl committed:
Update homeassistant/components/watchyourlan/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

drwahl committed:
chore: Refactor WatchYourLANConfigFlow to handle connection errors and unknown exceptions

drwahl committed:
chore: Update default manufacturer in WatchYourLANGenericSensor

drwahl committed:
chore: Remove mac_address from WatchYourLAN sensor entity descriptions

drwahl committed:
chore: Update WatchYourLAN sensor device info to use MAC address as connection

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,60 @@
+                errors["base"] = "cannot_connect"
+            if not errors:
+                return self.async_create_entry(
+                    title="WatchYourLAN", data={"url": user_input[CONF_URL]}
```
```suggestion
                    title="WatchYourLAN", data=user_input}
```

joostlek commented on homeassistant/components/watchyourlan/coordinator.py:
```diff
@@ -0,0 +1,46 @@
+        )
+
+        # Initialize the WatchYourLANClient with the base URL from the config
+        self.api_client = WatchYourLANClient(
+            base_url=self.config_entry.data[CONF_URL], async_mode=True
+        )
```
don't we have to set the `verify_ssl` here?

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,140 @@
+    async_add_entities(entities)
+
+
+class WatchYourLANOnlineStatusBinarySensor(
```
This is still relevant and should be removed

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,140 @@
+
+from .coordinator import WatchYourLANUpdateCoordinator
+
+WatchYourLANConfigEntry = ConfigEntry[WatchYourLANUpdateCoordinator]
```
Please import it from `__init__.py`

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,140 @@
+from .coordinator import WatchYourLANUpdateCoordinator
+
+WatchYourLANConfigEntry = ConfigEntry[WatchYourLANUpdateCoordinator]
+
+
+# Define entity descriptions for each sensor type
+ENTITY_DESCRIPTIONS = [
+    SensorEntityDescription(
+        key="online_status",
+        translation_key="online_status",
+    ),
+    SensorEntityDescription(
+        key="ip_address",
+        translation_key="ip_address",
+    ),
+    SensorEntityDescription(
+        key="iface",
+        translation_key="iface",
+    ),
+]
+
+# Define a mapping of sensor keys to device fields
+DEVICE_FIELD_MAPPING = {
+    "online_status": "Now",
+    "ip_address": "IP",
+    "iface": "Iface",
+}
```
```suggestion
ENTITY_DESCRIPTIONS = [
    SensorEntityDescription(
        key="IP",
        translation_key="ip_address",
    ),
    SensorEntityDescription(
        key="Iface",
        translation_key="iface",
    ),
]
```
Why don't you just use the `key` as the field to get from the coordinator data

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,140 @@
+        )
+
+    @property
+    def native_value(self) -> str:
+        """Return the native value of the sensor based on its description."""
+        device_field = DEVICE_FIELD_MAPPING.get(self.entity_description.key, None)
+
+        # Ensure that device_field is not None before accessing it
+        if device_field is not None:
+            return self.device.get(device_field, "Unknown")
+        return "Unknown"
```
return `None` instead of `"Unknown"`

joostlek commented on homeassistant/components/watchyourlan/strings.json:
```diff
@@ -0,0 +1,35 @@
+    "step": {
+      "user": {
+        "data": {
+          "url": "[%key:common::config_flow::data::host%]"
```
I am missing verify_ssl

joostlek commented on homeassistant/components/watchyourlan/strings.json:
```diff
@@ -0,0 +1,35 @@
+      "ip_address": {
+        "name": "IP Address"
+      },
+      "mac_address": {
+        "name": "MAC Address"
+      },
```
unused

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,124 @@
+    assert result["errors"] == {}
+
+    # Patch the API call to simulate valid response
+    with patch(
+        "homeassistant.components.watchyourlan.config_flow.WatchYourLANClient.get_all_hosts",
+        new_callable=AsyncMock,
+        return_value={"title": "WatchYourLAN", "url": "http://127.0.0.1:8840"},
+    ):
```
Consider creating a single mock for both the normal code as the config flow code. Checkout the `mock_airgradient_client` in `airgradient`

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,124 @@
+    )
+
+    # Check that the form is shown with no errors initially
+    assert result["type"] == FlowResultType.FORM
```
```suggestion
    assert result["type"] is FlowResultType.FORM
```

joostlek replied:
Please update this for the rest of the resutlt type checks as well

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,124 @@
+    """Test we get the form and create entry with valid data."""
+    # Initiate the flow
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

joostlek replied:
Please update this for the rest as well

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,124 @@
+        return_value={"title": "WatchYourLAN", "url": "http://127.0.0.1:8840"},
+    ):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                CONF_URL: "http://127.0.0.1:8840",
+                CONF_SSL: False,
+            },
+        )
+        await hass.async_block_till_done()
+
+    # Ensure the entry is created after retrying
+    assert result["type"] == FlowResultType.CREATE_ENTRY
+    assert result["title"] == "WatchYourLAN"
+    assert result["data"] == {
+        CONF_URL: "http://127.0.0.1:8840",
+    }
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_invalid_host(
+    hass: HomeAssistant, mock_setup_entry: AsyncMock
+) -> None:
+    """Test we handle an invalid host during the setup."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    # Simulate an invalid hostname or connection issue
+    with patch(
+        "homeassistant.components.watchyourlan.config_flow.WatchYourLANClient.get_all_hosts",
+        new_callable=AsyncMock,
+        side_effect=ConnectError("test"),
+    ):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                CONF_URL: "http://invalid-url",
+                CONF_SSL: False,
+            },
```
Please look into parametrizing the tests

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
I am missing a generic exception test and the test where we try to add the same configuration again

joostlek commented on tests/components/watchyourlan/test_sensor.py:
```diff
@@ -0,0 +1,215 @@
+    ]
+
+    # Mock the WatchYourLANUpdateCoordinator to return the devices without trying to fetch data
+    with (
+        patch(
+            "homeassistant.components.watchyourlan.sensor.WatchYourLANUpdateCoordinator._async_update_data",
+            return_value=devices,
+        ),
+        patch(
+            "homeassistant.components.watchyourlan.sensor.WatchYourLANUpdateCoordinator._schedule_refresh",
+            return_value=None,
+        ),
+    ):
```
we should not patch internal functions, we should patch the library to return what they should in a real life example

drwahl committed:
refactor: Update WatchYourLAN sensor entity descriptions and remove unnecessary code

This commit refactors the WatchYourLAN sensor entity descriptions by removing the mac_address field and simplifying the code. The native_value method in the WatchYourLANGenericSensor class has also been updated to return None when the device_field is not found. This improves the overall readability and maintainability of the code.

drwahl committed:
refactor: Remove mac_address from WatchYourLAN sensor entity descriptions

drwahl committed:
refactor: Update WatchYourLAN sensor entity descriptions and remove unnecessary code

drwahl committed:
refactor: Update WatchYourLAN sensor entity descriptions and remove unnecessary code

drwahl committed:
refactor: Remove commented out code for WatchYourLAN sensor setup

drwahl committed:
refactor: Update WatchYourLAN sensor entity descriptions and add verify_ssl field

Update the WatchYourLAN sensor entity descriptions in the strings.json file. Also, add the verify_ssl field to the user data step in the config flow.

autinerd commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,75 @@
+    """Set up the WatchYourLAN sensors."""
+    coordinator: WatchYourLANUpdateCoordinator = entry.runtime_data
+
+    entities: Sequence[SensorEntity | BinarySensorEntity] = [
```
```suggestion
    entities = [
```
This typing should not be needed.

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+        errors = {}
+
+        if user_input is not None:
+            # Set unique ID (use the URL as the unique ID)
+            await self.async_set_unique_id(user_input[CONF_URL])
+            # Abort if the unique ID is already configured
+            self._abort_if_unique_id_configured()
```
the url is a bad unique_id as it can change, please use `self._abort_entries_match` instead and don't set a unique id

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+
+            try:
+                hosts = await api_client.get_all_hosts()
+                if not hosts:
+                    errors["base"] = "cannot_connect"
```
move this out of the try block

joostlek commented on homeassistant/components/watchyourlan/coordinator.py:
```diff
@@ -0,0 +1,48 @@
+        self.api_client = WatchYourLANClient(
+            base_url=self.config_entry.data[CONF_URL],
+            async_mode=True,
+            verify_ssl=self.config_entry.data.get(CONF_VERIFY_SSL, True),
```
```suggestion
            verify_ssl=self.config_entry.data[CONF_VERIFY_SSL],
```
It's always set

joostlek commented on homeassistant/components/watchyourlan/icons.json:
```diff
@@ -0,0 +1,15 @@
+{
+  "entity": {
+    "sensor": {
+      "online_status": {
+        "default": "mdi:lan-connect"
+      },
```
Can be removed

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,75 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the WatchYourLAN sensors."""
+    coordinator: WatchYourLANUpdateCoordinator = entry.runtime_data
```
```suggestion
    coordinator = entry.runtime_data
```

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,75 @@
+    """Set up the WatchYourLAN sensors."""
+    coordinator: WatchYourLANUpdateCoordinator = entry.runtime_data
+
+    entities: Sequence[SensorEntity | BinarySensorEntity] = [
+        WatchYourLANGenericSensor(coordinator, device, description)
+        for device in coordinator.data
+        for description in ENTITY_DESCRIPTIONS
+    ]
+
+    async_add_entities(entities)
```
```suggestion
    async_add_entities(
        WatchYourLANGenericSensor(coordinator, device, description)
        for device in coordinator.data
        for description in ENTITY_DESCRIPTIONS
    )
```

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,75 @@
+                self.device.get("Name")
+                or f"WatchYourLAN {self.device.get('ID', 'Unknown')}"
+            ),
+            default_manufacturer=self.device.get("Hw", "Unknown Manufacturer"),
```
```suggestion
            default_manufacturer=self.device.get("Hw"),
```

drwahl replied:
mypy doesn't like this. It seems to want some default value:

```
homeassistant/components/watchyourlan/sensor.py:64: error: Incompatible types (expression has type "Any | None", TypedDict item "default_manufacturer" has type "str")  [typeddict-item]
Found 1 error in 1 file (checked 3 source files)
```

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,75 @@
+    def native_value(self) -> str | None:
+        """Return the native value of the sensor based on its description."""
+
+        return self.device.get(self.entity_description.key, None)
```
```suggestion
        return self.device.get(self.entity_description.key)
```

joostlek commented on homeassistant/components/watchyourlan/strings.json:
```diff
@@ -0,0 +1,33 @@
+  },
+  "entity": {
+    "sensor": {
+      "online_status": {
+        "name": "Online Status"
+      },
```
```suggestion
```

joostlek commented on homeassistant/components/watchyourlan/strings.json:
```diff
@@ -0,0 +1,33 @@
+    "step": {
+      "user": {
+        "data": {
+          "url": "[%key:common::config_flow::data::host%]",
+          "verify_ssl": "[%key:common::config_flow::data::verify_ssl%]"
```
please use `data_description` to give context about the fields

joostlek commented on homeassistant/components/watchyourlan/strings.json:
```diff
@@ -0,0 +1,33 @@
+      "online_status": {
+        "name": "Online Status"
+      },
+      "ip_address": {
+        "name": "IP Address"
+      },
+      "iface": {
+        "name": "Network Interface"
+      }
```
please use `snake_case`

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,149 @@
+    assert result["type"] is expected_result
+    assert result.get("errors", {}) == expected_errors
+
+    if expected_result is FlowResultType.CREATE_ENTRY:
```
We don't want if statement in the code, so I strongly recommend splitting the happy flow from the normal flow

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,149 @@
+        result["flow_id"],
+        user_input,
+    )
+    await hass.async_block_till_done()
```
```suggestion
```

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,149 @@
+        assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_generic_exception(
+    hass: HomeAssistant,
+    mock_watchyourlan_client: AsyncMock,
+    mock_setup_entry: AsyncMock,
+) -> None:
+    """Test handling an unexpected generic exception."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}
+    )
+
+    # Simulate an unexpected generic exception during validation
+    mock_watchyourlan_client.get_all_hosts.side_effect = Exception("Unexpected error")
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {CONF_URL: "http://127.0.0.1:8840", CONF_VERIFY_SSL: False},
+    )
+
+    # Ensure the form shows the unknown error
+    assert result["type"] is FlowResultType.FORM
+    assert result["errors"] == {"base": "unknown"}
```
Why didn't you parametrize this test?

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,149 @@
+    # Create a mock config entry to simulate an existing entry
+    mock_entry = MockConfigEntry(
+        domain=DOMAIN,
+        data={CONF_URL: "http://127.0.0.1:8840"},
```
Missing verify_ssl

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,149 @@
+    """Test if an already configured entry is detected and aborted."""
+
+    # Create a mock config entry to simulate an existing entry
+    mock_entry = MockConfigEntry(
```
Please create a separate fixture for the config entry, you can reuse this in the test_sensor.py

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,149 @@
+from tests.common import MockConfigEntry  # Import MockConfigEntry
+
+
+@pytest.fixture(name="mock_watchyourlan_client")
+def mock_watchyourlan_client_fixture():
+    """Fixture to mock WatchYourLANClient."""
+    with patch(
+        "homeassistant.components.watchyourlan.config_flow.WatchYourLANClient",
+    ) as mock_client:
+        instance = mock_client.return_value
+        instance.get_all_hosts = AsyncMock(return_value={})
+        yield instance
```
Move this to `conftest.py` and also patch the normal WatchYourLANClient. Check `airgradient` for an example

drwahl committed:
refactor: Update WatchYourLAN sensor entity descriptions and remove unnecessary code

drwahl commented:
Apologies for the slow response time here. Life events got in the way of me responding to your last feedback in a timely manner.

joostlek commented:
> Apologies for the slow response time here. Life events got in the way

Don't apologize for that, life events go before an integration :)

joostlek commented on homeassistant/components/watchyourlan/__init__.py:
```diff
@@ -0,0 +1,31 @@
+) -> bool:
+    """Set up WatchYourLAN from a config entry."""
+
+    # Create and store the coordinator, passing the entire ConfigEntry
+    coordinator = WatchYourLANUpdateCoordinator(hass)
+    await coordinator.async_config_entry_first_refresh()
+    entry.runtime_data = coordinator
+
+    # If the setup is successful, forward the entry to other platforms (e.g., sensor)
```
```suggestion
    coordinator = WatchYourLANUpdateCoordinator(hass)
    await coordinator.async_config_entry_first_refresh()
    entry.runtime_data = coordinator

```

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+        errors = {}
+
+        if user_input is not None:
+            # Abort if the unique ID is already configured
```
```suggestion
```

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+                _LOGGER.exception("Unexpected error during WatchYourLAN setup")
+                errors["base"] = "unknown"
+
+            if not hosts and not errors.get("base"):
```
```suggestion
            if not hosts and not errors:
```

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,71 @@
+    def native_value(self) -> str | None:
+        """Return the native value of the sensor based on its description."""
+
+        return self.device.get(self.entity_description.key, None)
```
```suggestion
        return self.device.get(self.entity_description.key)
```

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,71 @@
+                self.device.get("Name")
+                or f"WatchYourLAN {self.device.get('ID', 'Unknown')}"
+            ),
+            default_manufacturer=self.device.get("Hw", "Unknown manufacturer"),
```
```suggestion
            default_manufacturer=self.device.get("Hw"),
```
Rather have none than unknown

drwahl replied:
This patch makes mypy unhappy with the following error:
homeassistant/components/watchyourlan/sensor.py:64: error: Incompatible types (expression has type "Any | None", TypedDict item "default_manufacturer" has type "str")  [typeddict-item]

I'm not sure how to make mypy happy without setting a default value here. Any ideas?

joostlek replied:
Maybe we should move it outside of the definition and just only update the field of the device info object if we know it

joostlek commented on homeassistant/components/watchyourlan/strings.json:
```diff
@@ -0,0 +1,31 @@
+  "entity": {
+    "sensor": {
+      "ip_address": {
+        "name": "IP Address"
+      },
+      "iface": {
+        "name": "Network Interface"
```
```suggestion
        "name": "IP address"
      },
      "iface": {
        "name": "Network interface"
```

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,157 @@
+    [
+        (
+            {CONF_URL: "http://127.0.0.1:8840", CONF_VERIFY_SSL: False},
+            {"title": "WatchYourLAN", "url": "http://127.0.0.1:8840"},
```
Title is not a valid input

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,157 @@
+
+    # Ensure the form shows the expected errors
+    assert result["type"] is FlowResultType.FORM
+    assert result.get("errors", {}) == expected_errors
```
```suggestion
    assert result["errors"] == expected_errors
```

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,157 @@
+    )
+
+    # Ensure the form shows the expected errors
+    assert result["type"] is FlowResultType.FORM
```
All tests should end in either CREATE_ENTRY or ABORT. So please repatch the mock to work and let the flow finish. This way we test that the flow is able to succeed after an error

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,157 @@
+            ConnectError("test"),
+            {"base": "cannot_connect"},
+        ),
+        (
+            {CONF_URL: "http://invalid-url", CONF_VERIFY_SSL: False},
+            ConnectError("test"),
+            {"base": "cannot_connect"},
+        ),
```
The input isn't the cause on why the error is raised. I think it would be nice to combine this with the test below with the `Exception`.

joostlek commented on tests/components/watchyourlan/test_config_flow.py:
```diff
@@ -0,0 +1,157 @@
+    assert result.get("errors", {}) == expected_errors
+
+
+@pytest.mark.parametrize(
+    ("side_effect", "expected_error"),
+    [
+        (Exception("TEST Unexpected error"), {"base": "unknown"}),
+        (RuntimeError("TEST Runtime error"), {"base": "unknown"}),
+        (ValueError("TEST Value error"), {"base": "unknown"}),
+    ],
+)
+async def test_form_generic_exception(
+    hass: HomeAssistant,
+    mock_watchyourlan_client: AsyncMock,
```
Like just said, we only really need to have 1 test case, and I think we can merge this with the previous test

drwahl committed:
refactor: Update WatchYourLAN sensor entity descriptions and remove unnecessary code

drwahl committed:
Merge branch 'dev' into watchyourlan

drwahl committed:
Merge branch 'dev' into watchyourlan

joostlek commented:
I'll try to take a look at it again soon. Also please be aware that we now implemented a new quality scale that requires you to have a `quality_scale.yaml` file in your integration. I think you already check of a lot of boxes but we still want to add it :D

joostlek commented on homeassistant/components/watchyourlan/__init__.py:
```diff
@@ -0,0 +1,29 @@
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
```suggestion
async def async_unload_entry(hass: HomeAssistant, entry: WatchYourLANConfigEntry) -> bool:
```

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,67 @@
+        errors = {}
+
+        if user_input is not None:
+            self._async_abort_entries_match()
```
```suggestion
            self._async_abort_entries_match({})
```
We need to add what we are searching for

joostlek commented on homeassistant/components/watchyourlan/config_flow.py:
```diff
@@ -0,0 +1,67 @@
+                _LOGGER.exception("Unexpected error during WatchYourLAN setup")
+                errors["base"] = "unknown"
+
+            if not hosts and not errors:
```
Also, if hosts is an empty list, does that always mean something went wrong or does that just mean that there are no hosts connected?

joostlek commented on homeassistant/components/watchyourlan/sensor.py:
```diff
@@ -0,0 +1,71 @@
+                self.device.get("Name")
+                or f"WatchYourLAN {self.device.get('ID', 'Unknown')}"
+            ),
+            default_manufacturer=self.device.get("Hw", None),
```
```suggestion
            default_manufacturer=self.device.get("Hw"),
```

joostlek commented on homeassistant/components/watchyourlan/strings.json:
```diff
@@ -0,0 +1,31 @@
+        "data": {
+          "url": "[%key:common::config_flow::data::host%]",
+          "verify_ssl": "[%key:common::config_flow::data::verify_ssl%]",
+          "data_desctipion": "Enter the IP address of the WatchYourLAN server"
```
That's not the right way to add a data_description

joostlek commented on homeassistant/components/watchyourlan/strings.json:
```diff
@@ -0,0 +1,31 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```
it's a service, not a device

joostlek commented on tests/components/watchyourlan/conftest.py:
```diff
@@ -0,0 +1,74 @@
+        yield mock_setup_entry
+
+
+@pytest.fixture
+async def mock_watchyourlan_client_fixture() -> AsyncMock:
+    """Fixture to mock WatchYourLANClient."""
+    with patch(
+        "homeassistant.components.watchyourlan.config_flow.WatchYourLANClient",
+    ) as mock_client:
+        instance = mock_client.return_value
+        instance.get_all_hosts = AsyncMock(return_value={})
+        yield instance
```
Why don't you use this fixture as well in test_sensor.py if you also add the other patch?

joostlek replied:
Check airgradient for an example

joostlek commented on tests/components/watchyourlan/conftest.py:
```diff
@@ -0,0 +1,74 @@
+        "homeassistant.components.watchyourlan.config_flow.WatchYourLANClient",
+    ) as mock_client:
+        instance = mock_client.return_value
+        instance.get_all_hosts = AsyncMock(return_value={})
```
Why don't we always return something here by default? That is how 90% of all tests will work

joostlek commented on tests/components/watchyourlan/test_sensor.py:
```diff
@@ -0,0 +1,62 @@
+from homeassistant.core import HomeAssistant
+
+
+async def test_setup_sensors(hass: HomeAssistant, mock_config_entry) -> None:
```
Consider using snapshot tests. If you use `snapshot_platform`, you can execute `pytest ./tests/components/watchyourlan --snapshot-update` and it will generate a `.ambr` file for you with all the data structures fixated

joostlek reviewed (changes requested):
Like said before, this integration requires a quality_scale.yaml
